### PR TITLE
sbt-maven-resolver: Use Aether for resolving maven dependencies

### LINKED ISSUE #321: SBT doesn't see new SNAPSHOT dependency versions in local maven repos
I'm using SBT with a local maven repo cache (e.g. `~/.m2/repository`) for a Scala project that depends on a snapshot version of Java library that I develop at the same time. E.g. I add/change something in the java library and use maven to build a new snapshot version of the library. However, no combination of sbt commands (`update`, `clean`, etc.) will make it see the new snapshot version in the local repo - I have to delete the library from the ivy cache every time. It would be great if sbt/ivy would consider the timestamp of the snapshot version and update accordingly.

### LINKED ISSUE #647: Property in pom's version field results to wrong dependency resolution
(I'm going to edit this report so it's easier to review it @eed3si9n)
## steps

Create the following directory structure:

```
+- sbtpomtest/
    +- fakebase/
    +- fakeivy/
    +- sample/
        +- project/
            +- build.properties
        +- build.sbt
        +- sbt-launcher-script.sh
```

Here's `build.properties`:

```
sbt.version=0.13.0
```

Here's `build.sbt`:

``` scala
name := "sbtpomtest"

organization := "com.example"

version := "0.1-SNAPSHOT"

libraryDependencies ++= Seq(
   "org.jboss.shrinkwrap" % "shrinkwrap-api" % "1.0.1",
   "org.jboss.shrinkwrap" % "shrinkwrap-impl-base" % "1.0.1",
   "io.escalante" % "escalante-dist" % "0.2.0" artifacts(Artifact("escalante-dist", "zip", "zip")),
   "org.jboss.as" % "jboss-as-arquillian-container-embedded" % "7.x.incremental.546",
   "org.jboss.arquillian.core" % "arquillian-core-api" % "1.0.3.Final",
   "org.jboss.arquillian.core" % "arquillian-core-spi" % "1.0.3.Final",
   "org.jboss.arquillian.core" % "arquillian-core-impl-base" % "1.0.3.Final",
   "org.jboss.arquillian.container" % "arquillian-container-spi" % "1.0.3.Final",
   "org.jboss.arquillian.container" % "arquillian-container-impl-base" % "1.0.3.Final",
   "org.jboss.arquillian.test" % "arquillian-test-impl-base" % "1.0.3.Final",
   "org.yaml" % "snakeyaml" % "1.8"
)

resolvers ++= Seq(
  "JBoss repository" at "http://repository.jboss.org/nexus/content/groups/public/",
  "Project Odd repository" at "http://repository-projectodd.forge.cloudbees.com/upstream/"
)
```

and here's `sbt-launcher-script.sh`:

```
#!/bin/sh
SBT_OPTS='-Dsbt.global.base=/Users/foo/sbtpomtest/fakebase -Dsbt.ivy.home=/Users/foo/sbtpomtest/fakeivy '$SBT_OPTS
exec java -Xmx512M ${SBT_OPTS} -XX:MaxPermSize=512M -jar ~/Applications/sbt/sbt-launch-0.13.0.jar "$@"
```

(The script needs to be changed depending on where your jar, `fakebase/`, and `fakeivy/` are)

`chmod +x` and run the script:

```
[sbtpomtest]$ cd sample
[sample]$ chmod +x ./sbt-launcher-script.sh 
[sample]$ ./sbt-launcher-script.sh
```

Next, run `compile` from the sbt shell.

```
> compile
[info] Updating {file:/Users/foo/sbtpomtest/sample/}sample...
```
## result (as of 12/2/2013 using sbt 0.13)

```
[info] downloading http://repo1.maven.org/maven2/org/jboss/shrinkwrap/resolver/shrinkwrap-resolver-api/1.0.0-beta-5/shrinkwrap-resolver-api-1.0.0-beta-5.jar ...
[info]  [SUCCESSFUL ] org.jboss.shrinkwrap.resolver#shrinkwrap-resolver-api;1.0.0-beta-5!shrinkwrap-resolver-api.jar (217ms)
[info] downloading http://repo1.maven.org/maven2/org/jboss/shrinkwrap/resolver/shrinkwrap-resolver-api-maven/1.0.0-beta-5/shrinkwrap-resolver-api-maven-1.0.0-beta-5.jar ...
[info]  [SUCCESSFUL ] org.jboss.shrinkwrap.resolver#shrinkwrap-resolver-api-maven;1.0.0-beta-5!shrinkwrap-resolver-api-maven.jar (209ms)
[info] downloading http://repo1.maven.org/maven2/org/apache/maven/wagon/wagon-file/1.0-beta-7/wagon-file-1.0-beta-7.jar ...
[info]  [SUCCESSFUL ] org.apache.maven.wagon#wagon-file;1.0-beta-7!wagon-file.jar (219ms)
[info] downloading http://repo1.maven.org/maven2/org/jboss/arquillian/test/arquillian-test-spi/1.0.3.Final/arquillian-test-spi-1.0.3.Final.jar ...
[info]  [SUCCESSFUL ] org.jboss.arquillian.test#arquillian-test-spi;1.0.3.Final!arquillian-test-spi.jar (221ms)
[info] Done updating.
[success] Total time: 475 s, completed Dec 2, 2013 10:40:13 PM
> 
```
## original bug report (was "Wrong dependency version resolution leading to build halt")

Hi, as a result #646 I've nuked my `~/.ivy2/local` and `~/.ivy2/cache` folders and when I try to update my SBT plugin I get:

```
[warn]  :: org.jboss.shrinkwrap.descriptors#shrinkwrap-descriptors-api-base;1.0.3.Final: not found
[warn]  :: org.jboss.shrinkwrap.descriptors#shrinkwrap-descriptors-spi;1.0.3.Final: not found
```

I think this is a problem where the dependency version to use for this dependency is not correctly calculated (not my lucky day today...). Let me explain:

Looking at the debug log I see: https://gist.github.com/4564402

That Shrinkwrap dependency comes from org.jboss.arquillian.config#arquillian-config-impl-base;1.0.3.Final dependency. The pom for this dependency is:

https://repository.jboss.org/nexus/content/groups/public/org/jboss/arquillian/config/arquillian-config-impl-base/1.0.3.Final/arquillian-config-impl-base-1.0.3.Final.pom

which states:

```
<dependency>
    <groupId>org.jboss.shrinkwrap.descriptors</groupId>
    <artifactId>shrinkwrap-descriptors-spi</artifactId>
    <scope>compile</scope>
</dependency>
```

This dependency's version should be resolved checking the parent, which is: 
https://repository.jboss.org/nexus/content/groups/public/org/jboss/arquillian/arquillian-build/1.0.3.Final/arquillian-build-1.0.3.Final.pom

This build parent defines:

```
<dependency>
    <groupId>org.jboss.shrinkwrap.descriptors</groupId>
    <artifactId>shrinkwrap-descriptors-bom</artifactId>
    <version>${version.shrinkwrap_descriptors}</version>
    <type>pom</type>
    <scope>import</scope>
</dependency>
```

`${version.shrinkwrap_descriptors}` is defined in this last pom's parent:
https://repository.jboss.org/nexus/content/groups/public/org/jboss/arquillian/arquillian-parent/1.0.3.Final/arquillian-parent-1.0.3.Final.pom

which defines:

<version.shrinkwrap_descriptors>2.0.0-alpha-3</version.shrinkwrap_descriptors>

And that references:

https://repository.jboss.org/nexus/content/groups/public/org/jboss/shrinkwrap/descriptors/shrinkwrap-descriptors-bom/2.0.0-alpha-3/shrinkwrap-descriptors-bom-2.0.0-alpha-3.pom

Which defines:

```
<dependency>
    <groupId>org.jboss.shrinkwrap.descriptors</groupId>
    <artifactId>shrinkwrap-descriptors-spi</artifactId>
    <version>${project.version}</version>
</dependency>
```

Where project.version=2.0.0-alpha-3, so that should resolve in:

https://repository.jboss.org/nexus/content/groups/public/org/jboss/shrinkwrap/descriptors/shrinkwrap-descriptors-spi/2.0.0-alpha-3/shrinkwrap-descriptors-spi-2.0.0-alpha-3.jar

What is going on here? Is this an SBT or ivy issue?

### LINKED ISSUE #1322: SBT doesn't honor Maven's uniqueVersions
## Steps
- Publish an `SNAPSHOT` artifact to a Maven repository (with `publishMavenStyle` set to `true`, obviously)
- Try to fetch this artifact as a dependency from Maven 3
- Notice the trouble you're into when trying to fetch the latest `SNAPSHOT`...
## Problems

Maven's support for "unique versions" was deprecated in Maven 2, which reached EOL in February 2014, and definitely disabled in Maven 3. 
However, SBT publishes `SNAPSHOT` artifacts as "unique versions", which causes all kinds of trouble when trying to fetch dependencies from Maven.

Of course, I know SBT isn't Maven and does not try to be Maven, which is great, I love SBT for that, among other things.
However, I think that SBT can't ignore Maven either, as it does already with the possibility to publish POMs, to publish to Maven repos, etc...
Allowing to publish "non-unique versions" is just another step ;)

I know that the [sbt-unique-version plugin](https://github.com/sbt/sbt-unique-version) allows to do that, but I honestly think this should be standard behaviour when publishing `SNAPSHOT` dependencies to a remote Maven repository (as long as it doesn't cause problems somewhere else, of course).

[The sbt-aether-deploy plugin](https://github.com/arktekk/sbt-aether-deploy) fixes that issue too, but pulling Maven in SBT to publish artifacts is a tiny bit overkill...
Plus, there is a version conflict on `httpclient` (both used by SBT and the plugin, pulled by the dependency on Aether) with SBT 0.13, causing the plugin to fail to publish artifacts (SBT isn't to blame for that, of course).

As the problem only arises when trying to fetch a dependency from a remote Maven repository (not a local repo), I suggest that : 
- Publishing to a remote repository (eg. using the `publish` task) 
- With `publishMavenStyle` to `true`

could trigger the change in the artifact name to include the timestamp.

Should you agree with this, I could provide a PR if you like.

### LINKED ISSUE #1616: Maven local resolver with parent POM
Somewhere between 0.13.5 and 0.13.6, the Maven local resolver stopped being able to resolve parent POM's from Maven projects.

If I have a parent pom:

``` xml
<?xml version="1.0"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.example</groupId>
    <artifactId>example-parent</artifactId>
    <version>1.0-SNAPSHOT</version>
    <packaging>pom</packaging>

</project>
```

a child pom:

``` xml
<?xml version="1.0"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <parent>
        <groupId>com.example</groupId>
        <artifactId>example-parent</artifactId>
        <version>1.0-SNAPSHOT</version>
    </parent>

    <groupId>com.example</groupId>
    <artifactId>example-child</artifactId>
    <version>1.0-SNAPSHOT</version>
    <packaging>jar</packaging>

</project>
```

both of which are installed to my local M2 repo using `mvn install`. The following build.sbt works in 0.13.5:

```
name := "example-sbt"

version := "1.0-SNAPSHOT"

libraryDependencies += "com.example" % "example-child" % "1.0-SNAPSHOT"

resolvers += Resolver.mavenLocal
```

but fails in 0.13.6 with the error:

```
<snip>

[info] Updating {file:/Users/csimpson/Desktop/sbt-mvn-parent/sbt/}sbt...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: com.example#example-child;1.0-SNAPSHOT:  Maven2 Local: no ivy file nor artifact found for com.example#example-parent;1.0-SNAPSHOT
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]      com.example:example-child:1.0-SNAPSHOT (/Users/csimpson/Desktop/sbt-mvn-parent/sbt/build.sbt#L5-6)
[warn]        +- example-sbt:example-sbt_2.10:1.0-SNAPSHOT
sbt.ResolveException: unresolved dependency: com.example#example-child;1.0-SNAPSHOT:    Maven2 Local: no ivy file nor artifact found for com.example#example-parent;1.0-SNAPSHOT

<snip>
```

It might be an issue with Ivy caching, since if I run SBT 0.13.5 first, then 0.13.6 will succeed, but if I delete `~/.ivy2/cache/com.example`, then 0.13.6 fails to resolve again. I've verified that Maven is able to resolve the parent POM when depending on the child and that SBT is able to resolve Maven artifacts that don't have a parent POM.

### PR DESCRIPTION
**Pull request by jsuereth**
sbt 0.13.8 adds an extension point in the dependency resolution to customize Maven resolvers.
This allows us to write sbt-maven-resolver auto plugin, which internally uses Eclipse Aether
to resolve Maven dependencies instead of Apache Ivy.

To enable this plugin, add the following to `project/maven.sbt` (or `project/plugin.sbt` the file name doesn't matter):

``` scala
libraryDependencies += Defaults.sbtPluginExtra("org.scala-sbt" % "sbt-maven-resolver" % sbtVersion.value,
  sbtBinaryVersion.value, scalaBinaryVersion.value)
```

This will create a new `~/.ivy2/maven-cache` directory, which contains the Aether cache of files.
You may notice some file will be re-downloaded for the new cache layout.
Additionally, sbt will now be able to fully construct
`maven-metadata.xml` files when publishing to remote repositories or when publishing to the local `~/.m2/repository`.
This should help erase many of the deficiencies encountered when using Maven and sbt together.

**Notes and known limitations**:
- sbt-maven-resolver requires sbt 0.13.8 and above.
- The current implementation does not support Ivy-style dynamic revisions, such as "2.10.+" or "latest.snapshot".  This
  is a fixable situation, but the version range query and Ivy -> Maven version range translation code has not been migrated.
## original case

This change adds the usage of Aether to replace Ivy's default maven dependency handling.
- Creates a new type of Ivy dependency resolver which resolves dependencies using Aether
- Wires Aether's artifact download mechanism to use Ivy's download mechanisms
- Adds configuration/adaptation to Aether so it can successfully resolve sbt plugins
  <s>\- Creates a new `UpdateOptions` flag which enables/disables Aether resolution. (Right now defaults to true so we can see if TravisCI passe swith Aether resolution, then we'll add a commit to default to false to ensure we didn't break previous behavior).</s>
- Updates tests which rely on Aether behavior for correctness to default to aether resolution.
## Known bugs fixed:

Fixes #1322
Fixes #321
Fixes #647
Fixes #1616
## Possible bugs fixed (hoping to verify soon):
#1676
#679

Review by @eed3si9n 

NOTE: Please do not merge until this notice is taken down, which implies that Aether resolution is no longer the default (keeping existing behavior and having it opt-in) as well as ensuring that the plugin deduplication concern has a test in MavenResolutionSpec.

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -27,4 +27,41 @@ object Dependencies {
...
+  val aetherImpl = "org.eclipse.aether" % "aether-impl" % aetherVersion
+  val aetherUtil = "org.eclipse.aether" % "aether-util" % aetherVersion
+  val aetherTransportFile = "org.eclipse.aether" % "aether-transport-file" % aetherVersion
+  val aetherTransportWagon = "org.eclipse.aether" % "aether-transport-wagon" % aetherVersion
+  val aetherTransportHttp = "org.eclipse.aether" % "aether-transport-http" % aetherVersion
+  val aetherConnectorBasic = "org.eclipse.aether" % "aether-connector-basic" % aetherVersion
+  val sisuPlexus = ("org.eclipse.sisu" % "org.eclipse.sisu.plexus" % "0.3.0.M1").exclude("javax.enterprise", "cdi-api").exclude("com.google.code.findbugs", "jsr305")
+  val guice = "com.google.inject" % "guice" % "3.0"
+  val guava = "com.google.guava" % "guava" % "18.0"
+  val javaxInject = "javax.inject" % "javax.inject" % "1"
```
Comment (eed3si9n) [inline project/Dependencies.scala:48]: Are these all going to be part of the plugin classpath? Should we fork the jvm to isolate the classpath?

Comment (jsuereth) [inline project/Dependencies.scala:48]: Yes, it'll be part of the classpath.  (well, not all of those but a lot).   And I don't think JVM isolation of the classpath with a fork is a terribly great idea.

Comment (eed3si9n) [inline project/Dependencies.scala:48]: What about coding some scaffolding in chain resolver, but largely moving `MavenRemoteRepositoryResolver` code into a plugin? I am not sure if it's necessarily a good idea, but is that a viable approach?

Comment (jsuereth) [inline project/Dependencies.scala:48]: I dunno if it's viable.   The amount of public API I'd have to touch would be a lot larger.   Right now I was able to hide most of it in UpdateOptions.   If we don't have it in sbt, i'd have to expose something for the plugin to use, which will be a lot of machinery.

**Comment by jsuereth** [comment]
@eed3si9n The commits are now squashed into the best coherent split I could think of.

**Comment by dwijnand** [comment]
(Apologies if this is the wrong forum - please tell me)

AFAIK one of the differences between Aether and Ivy (as used in sbt) is version conflict resolution with Ivy defaulting to "highest version value" while Aether is "closest dependency node", which therefore can gives two different end results. With this pull request will this change for maven dependencies?

**Comment by jsuereth** [comment]
@dwijnand no.  Conflict resolution would still be done in Ivy.  The diffference is dependencyManagement sections and parent-poms wouldbe correctly read first.

**Comment by dwijnand** [comment]
Thank you Josh

**Comment by jsuereth** [comment]
Note to self ->

Updates to disambiguate plugins in the aether cache lead to breaking `project/transitive-plugins` scripted test.  I'll fix that shortly.

**Comment by jsuereth** [comment]
@eed3si9n If the new .travis.yml configuration succeeds, this one is good to merge

**Comment by eed3si9n** [comment]
Travis failed on one job. Restarting it.

**Comment by schlichtanders** [comment]
when trying to use http://central.maven.org/maven2/org/apache/maven/maven-plugin-api/3.3.9/maven-plugin-api-3.3.9.pom I unfortunately cannot resolve the maven property ``${project.version}`` for the subpackage ``maven-artifacts``  (the other subpackages seem to work fine).

here the dependency tag from the pom:
```
<dependencies>
<dependency>
<groupId>org.apache.maven</groupId>
<artifactId>maven-model</artifactId>
</dependency>
<dependency>
<groupId>org.apache.maven</groupId>
<artifactId>maven-artifact</artifactId>
<exclusions>
<exclusion>
<artifactId>wagon-provider-api</artifactId>
<groupId>org.apache.maven.wagon</groupId>
</exclusion>
</exclusions>
</dependency>
<dependency>
<groupId>org.eclipse.sisu</groupId>
<artifactId>org.eclipse.sisu.plexus</artifactId>
</dependency>
</dependencies>
```
The only difference I can see between ``maven-artifact`` and say ``maven-model`` is the ``exclusions`` tag. **Can it be that this is not properly handled by Aether?**


---------------------------

I really only added the suggested line to my project/plugin.sbt
```
libraryDependencies += Defaults.sbtPluginExtra("org.scala-sbt" % "sbt-maven-resolver" % sbtVersion.value,
  sbtBinaryVersion.value, scalaBinaryVersion.value)
```

(in the final run I try to install ``addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.14.7")`` behind a proxy, however ran into a lot of these unresolved maven properties...
my sbt version is 0.13.7)

Any help is highly appreciated

**Pull request merged**